### ISSUE TITLE
backprop through torch.where backprops nans through path that was not taken

### ISSUE THREAD

**Issue by jonasrauber** [pytorch/pytorch#36923]
I think backprop through `torch.where` is wrong in certain special cases.

## To Reproduce

```python
t = 0
x = torch.ones(()).requires_grad_()
y = t * (x / t)  # just an example; anything that produces nan's works
z = torch.where(x >= t, x, y)
z.backward()

# the forward pass works fine (the `nan`'s in `y` do not affect z)
# NOTE: this is unlike a naive implement of where that does `cond * x + (1 - cond) * y`
print(z)
# tensor(1., grad_fn=<SWhereBackward>)

# but the backward pass backprops the `nan`'s from y into x, even though the y path is never taken in torch.where
print(x.grad)
# tensor(nan)
```

## Expected behavior
```python
print(x.grad)
# tensor(1.)
```
This would be the correct gradient.

In practice, this bug can easily happen if one runs the above code for different t (including 0) and assumes that the `nan`'s for `t = 0` should not matter because the `torch.where` always selects the first path that has no `nan`s (and the forward pass does handle it correctly).
## Environment

 - PyTorch Version 1.4

**Comment by albanD** [comment]
Hi,

This is a duplicate of https://github.com/pytorch/pytorch/issues/23395
Closing in favor of the other issue that explain why it happens.
